### PR TITLE
Add subscription hooks to Mercurius

### PR DESCRIPTION
**Pull request by jonnydgreen**
Closes: https://github.com/mercurius-js/mercurius/issues/406

### CODE REVIEW

**Review by mcollina — COMMENTED** [review]
Good work! Why did you add all the nextTick?

**Comment by jonnydgreen** [comment]
Thanks! For `nextTick`, I followed the Node.js docs examples that used `once` as I wanted to guarantee that I was listening with `once` before the event was fired: https://nodejs.org/api/events.html#events_events_once_emitter_name_options

File: test/subscription-hooks.js [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+      type: 'connection_init'
+    }))
+    client.write(JSON.stringify({
+      id: 1,
+      type: 'start',
+      payload: {
+        query
+      }
+    }))
+  })
```
Comment (mcollina) [inline test/subscription-hooks.js:150]: there is no need for this `nextTick`.

File: test/subscription-hooks.js [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+
+  {
+    const [chunk] = await once(client, 'data')
+    const data = JSON.parse(chunk)
+    t.is(data.type, 'connection_ack')
+  }
+
+  process.nextTick(() => {
+    sendTestMutation(app)
+  })
```
Comment (mcollina) [inline test/subscription-hooks.js:160]: there is no need for this nextTick

**Review by mcollina — CHANGES_REQUESTED** [review]
I think you can safely remove most (if not all) `nextTick`. When you write to a socket the messages go through a few full event loop cycles.

**Comment by jonnydgreen** [comment]
Makes sense - I'll get on that now!

File: lib/subscription-connection.js [hunk 3]
```diff
@@ -196,6 +206,15 @@ module.exports = class SubscriptionConnection {
 
     // TODO implement backpressure
     for await (const value of subIter) {
+      // Trigger onSubscriptionResolution hook
+      if (this.context.onSubscriptionResolution !== null) {
+        try {
+          await onSubscriptionResolutionHandler({ execution: value, context })
+        } catch (error) {
+          this.fastify.log.error(error)
+          this.handleConnectionClose()
```
Comment (mcollina) [inline lib/subscription-connection.js:215]: I think we should return after calling this function.

**Review by mcollina — CHANGES_REQUESTED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Review by PacoDu — COMMENTED** [review]
lgtm ! Maybe we could also add a hook when the subscription ends ? (when `GQL_STOP` is received)

**Comment by jonnydgreen** [comment]
For sure - I'll add that later along with the updated lifecycle! :)

**Review by PacoDu — APPROVED** [review]
lgtm !

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**